### PR TITLE
JavaScript overhauled and ready for versioning

### PR DESCRIPTION
**Pull request by pgilmorepf**

### CODE REVIEW

File: PlayFabApiTest.js [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+            else {
+                // Confirm the successful login
+                MandatoryLoginCallback(result, error)
+            }
+        };
+        var RegisterCallback = function (result, error) {
+            // Second login MUST succeed
+            PlayFabApiTests.VerifyNullError(result, error, assert, "Testing Registration result");
+
+            // Log in again, this time with the newly registered account
```
Comment (zac-playfab) [inline PlayFabApiTest.js]: Unless you explicitly want to call this login feature, you can use the auth ticket returned by RegisterPlayFabUser

File: PlayFabApiTest.js [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        };
+        loginDone = assert.async();
+        PlayFabClientSDK.LoginWithEmailAddress(loginRequest, PlayFabApiTests.CallbackWrapper("OptionalLoginCallback", OptionalLoginCallback, assert));
+    },
+
+    UserDataApi: function (assert) {
+        var getDataRequest = {}; // null also works
+
+        // This test is always exactly 3 async calls
+        get1Done = assert.async();
```
Comment (zac-playfab) [inline PlayFabApiTest.js]: Can you think of a better var name for these?

File: PlayFabApiTest.js [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+        };
+
+        var GetLeaderboardCallback_C = function (result, error) {
+            PlayFabApiTests.VerifyNullError(result, error, assert, "Testing GetLeaderboard result");
+            if (result != null) {
+                assert.ok(result.Leaderboard != null, "Testing GetLeaderboard content");
+                assert.ok(result.Leaderboard.length > 0, "Testing GetLeaderboard content-length");
+            }
+
+            lbDone_C();
```
Comment (zac-playfab) [inline PlayFabApiTest.js]: var names?

**Comment by zac-playfab** [comment]
See comments, & LGTM

File: PlayFabApiTest.js (hunk 1 above)
Comment (pgilmorepf) [inline PlayFabApiTest.js]: OIC, yes.  I could do that.  I guess I fell into the pattern of: login, [register, login].  For now, I think we should leave it this way, as it matches all other auto-testing in all sdks.  Maybe I can make a pass later to remove the second login call everywhere, as you're right I don't think it's necessary.

**Comment by pgilmorepf** [comment]
Yeah, I agree with var names.  Probably a good idea to do a bit of naming-cleanup in a future pass.

**Pull request merged**